### PR TITLE
[RLlib] Learner group checkpointing

### PR DESCRIPTION
**Pull request by avnishn**
Signed-off-by: Avnish <avnishnarayan@gmail.com>

Implement multinode learner group checkpointing and tests.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/rl_module/rl_module.py [hunk 1]
```diff
@@ -609,3 +609,7 @@ def as_multi_agent(self) -> "MultiAgentRLModule":
         marl_module = MultiAgentRLModule()
         marl_module.add_module(DEFAULT_POLICY_ID, self)
         return marl_module
+
+    def unwrapped(self) -> "RLModule":
+        """Returns the underlying module if this module is a wrapper."""
```
Comment (sven1977) [inline rllib/core/rl_module/rl_module.py:614]: Can you specify, what wrapper here means?
Like what are examples for RLModule wrappers?

File: rllib/core/rl_module/tests/test_marl_module.py [hunk 2]
```diff
@@ -7,13 +7,11 @@
     MultiAgentRLModuleConfig,
 )
 from ray.rllib.core.testing.torch.bc_module import DiscreteBCTorchModule
+from ray.rllib.core.testing.utils import DEFAULT_POLICY_ID
 from ray.rllib.env.multi_agent_env import make_multi_agent
 from ray.rllib.utils.test_utils import check
 
 
-DEFAULT_POLICY_ID = "default_policy"
```
Comment (sven1977) [inline rllib/core/rl_module/tests/test_marl_module.py:14]: Nice! :)

File: rllib/core/rl_module/rl_module.py (hunk 1 above)
Comment (avnishn) [reply rllib/core/rl_module/rl_module.py:614]: torch rl modules get wrapped with the torch ddp rl module wrapper

File: rllib/core/testing/utils.py [hunk 3]
```diff
@@ -26,6 +25,9 @@
 Optimizer = Union["tf.keras.optimizers.Optimizer", "torch.optim.Optimizer"]
 
 
+DEFAULT_POLICY_ID = "default_policy"
```
Comment (sven1977) [inline rllib/core/testing/utils.py:28]: Why don't we import this from `policy.py` here?

Comment (avnishn) [reply rllib/core/testing/utils.py:28]: I want to avoid mixing policy code in the new stack.

File: rllib/core/learner/tf/tf_learner.py [hunk 4]
```diff
@@ -68,9 +68,7 @@ def __init__(
         super().__init__(framework_hyperparameters=framework_hyperparameters, **kwargs)
 
         self._enable_tf_function = framework_hyperparameters.eager_tracing
-        # the default strategy is a no-op that can be used in the local mode
-        # cpu only case, build will override this if needed.
-        self._strategy = tf.distribute.get_strategy()
+        self._strategy = None
```
Comment (sven1977) [inline rllib/core/learner/tf/tf_learner.py:71]: Can you leave the comment on what `self._strategy` is (or should be when not None)?

File: rllib/core/learner/tf/tf_learner.py [hunk 5]
```diff
@@ -349,6 +347,25 @@ def remove_module(self, module_id: ModuleID) -> None:
         if self._enable_tf_function:
             self._update_fn = tf.function(self._do_update_fn, reduce_retracing=True)
 
+    def _make_distributed_strategy(self):
+        """Create a distributed strategy for the learner."""
```
Comment (sven1977) [inline rllib/core/learner/tf/tf_learner.py:351]: Same, can you add a little more explanation here on what a "strategy" is and which types exist (an example?)?

File: rllib/core/learner/tf/tf_learner.py (hunk 4 above)
Comment (avnishn) [reply rllib/core/learner/tf/tf_learner.py:71]: Strategy is a tf distributed strategy object that is used for the ddp logic.

File: rllib/core/learner/tf/tf_learner.py (hunk 5 above)
Comment (avnishn) [reply rllib/core/learner/tf/tf_learner.py:351]: Strategy is a tf distributed strategy object. 

The different types of strategies are contained within the function.

File: rllib/core/rl_module/marl_module.py [hunk 6]
```diff
@@ -483,7 +483,7 @@ def from_module(self, module: MultiAgentRLModule) -> "MultiAgentRLModuleSpec":
             The MultiAgentRLModuleSpec.
         """
         module_specs = {
-            module_id: SingleAgentRLModuleSpec.from_module(rl_module)
+            module_id: SingleAgentRLModuleSpec.from_module(rl_module.unwrapped())
```
Comment (sven1977) [inline rllib/core/rl_module/marl_module.py:486]: Nit: Explain why we need to unwrap here. rl_module could be a framework-specific DDP wrapper?

File: rllib/core/learner/tests/test_learner_group.py [hunk 7]
```diff
@@ -19,7 +25,7 @@
 
 REMOTE_SCALING_CONFIGS = {
     "remote-cpu": LearnerGroupScalingConfig(num_workers=1),
-    "remote-gpu": LearnerGroupScalingConfig(num_workers=1, num_gpus_per_worker=0.5),
+    "remote-gpu": LearnerGroupScalingConfig(num_workers=1, num_gpus_per_worker=1),
```
Comment (sven1977) [inline rllib/core/learner/tests/test_learner_group.py:28]: Why did we change this? Would it break if we used fractional GPUs here?

File: rllib/core/learner/tests/test_learner_group.py [hunk 8]
```diff
@@ -268,6 +274,66 @@ def test_async_update(self):
...
+            del new_learner_group
+
+            # construct a new learner group and load the initial state of the learner
+            learner_group = get_learner_group(
+                fw, env, scaling_config, eager_tracing=True
+            )
+            learner_group.load_state(initial_learner_checkpoint_dir)
+            check(learner_group.get_weights(), initial_learner_group_weights)
+            learner_group.update(batch.as_multi_agent(), reduce_fn=None)
+            results_without_break = learner_group.update(
```
Comment (sven1977) [inline rllib/core/learner/tests/test_learner_group.py:328]: Could we check here again to see whether the weights after one update (based off the initial state) are the same as the weights of the original learner (after one update)?

**Review by sven1977 — COMMENTED** [review]
Awesome PR @avnish! Thanks for covering this important feature in our release tests from here on.
Just a few nits, questions, and suggestions for better comments.

File: rllib/core/learner/tests/test_learner_group.py (hunk 7 above)
Comment (avnishn) [reply rllib/core/learner/tests/test_learner_group.py:28]: this learner/group actually won't even take fractional gpus. so it is pointless. I changed it while I was doing some debugging

File: rllib/core/rl_module/rl_module.py (hunk 1 above)
Comment (avnishn) [reply rllib/core/rl_module/rl_module.py:614]: done

File: rllib/core/learner/tf/tf_learner.py (hunk 4 above)
Comment (avnishn) [reply rllib/core/learner/tf/tf_learner.py:71]: added a param notation.

Comment (avnishn) [reply rllib/core/learner/tf/tf_learner.py:71]: done

File: rllib/core/rl_module/marl_module.py (hunk 6 above)
Comment (avnishn) [reply rllib/core/rl_module/marl_module.py:486]: done.

**Review by sven1977 — APPROVED** [review]
LGTM now. Thanks!

File: rllib/core/learner/learner_group.py [hunk 9]
```diff
@@ -404,6 +408,134 @@ def set_is_module_trainable(
...
+                    worker_ip_addr, worker_temp_dir, self_ip_addr, path
+                )
+
+                # creating this function here instead of making it a member funciton
+                # becasue it uses the worker_temp_dir variable, and this can't
+                # be passed in as an argument to foreach_actor
+                def remove_dir(w):
+                    import shutil
+
+                    shutil.rmtree(worker_temp_dir)
```
Comment (gjoliver) [inline rllib/core/learner/learner_group.py:456]: can you make this a member function on Worker as well?
so you can do ``lambda w: w.remove_worker_temp_dir()`` below.

**Review by gjoliver — COMMENTED** [review]
let me talk to you offline about how you intend to use this?

File: rllib/core/learner/learner_group.py [hunk 10]
```diff
@@ -404,6 +408,134 @@ def set_is_module_trainable(
...
+        else:
+            head_node_ip = socket.gethostbyname(socket.gethostname())
+            workers = self._worker_manager.healthy_actor_ids()
+
+            def _load_state(w):
+                # doing imports here since they might not be imported on the worker
+                import socket
+                import tempfile
+
+                hostname = socket.gethostname()
```
Comment (avnishn) [inline rllib/core/learner/learner_group.py:490]: ray.util.get_node_ip

**Pull request merged**